crates/bevy_pbr/src/render/skin.rs

```diff
@@ -150,70 +277,334 @@ pub fn prepare_skins(
+                skinned_mesh_inverse_bindposes.get(joint_index)
+            });
+        let joint_matrix = match (skinned_mesh_inverse_bindpose, joints.get(joint)) {
+            (Some(skinned_mesh_inverse_bindpose), Ok(transform)) => {
+                transform.affine() * *skinned_mesh_inverse_bindpose
+            }
+            _ => Mat4::IDENTITY,
+        };
+
+        // Write in the new joint matrix, growing the staging buffer if
+        // necessary.
+        let buffer_index = skin_uniform_info.offset() as usize + joint_index;
+        if skin_uniforms.current_staging_buffer.len() < buffer_index + 1 {
+            skin_uniforms
+                .current_staging_buffer
+                .resize(buffer_index + 1, Mat4::IDENTITY);
         }
-        last_start = last_start.max(start);
+        skin_uniforms.current_staging_buffer[buffer_index] = joint_matrix;
+
+        // Record the inverse mapping from the joint back to the skin. We use
+        // this in order to perform fine-grained joint extraction.
+        skin_uniforms
+            .joint_to_skins
+            .entry(MainEntity::from(joint))
+            .or_default()
+            .push(skinned_mesh_entity);
+    }
+
+    // Record the number of joints.
+    skin_uniforms.total_joints += skinned_mesh.joints.len();
+
+    skin_uniforms
+        .skin_uniform_info
+        .insert(skinned_mesh_entity, skin_uniform_info);
+}
+
+/// Deallocates a skin and removes it from the [`SkinUniforms`].
+fn remove_skin(skin_uniforms: &mut SkinUniforms, skinned_mesh_entity: MainEntity) {
+    let Some(old_skin_uniform_info) = skin_uniforms.skin_uniform_info.remove(&skinned_mesh_entity)
+    else {
+        return;
+    };
 
-        // Pad to 256 byte alignment if we're using a uniform buffer.
-        // There's no need to do this if we're using storage buffers, though.
-        if skins_use_uniform_buffers {
-            while buffer.len() % 4 != 0 {
-                buffer.push(Mat4::ZERO);
+    // Free the allocation.
+    skin_uniforms
+        .allocator
+        .free(old_skin_uniform_info.allocation);
+
+    // Remove the inverse mapping from each joint back to the skin.
+    for &joint in &old_skin_uniform_info.joints {
+        if let Entry::Occupied(mut entry) = skin_uniforms.joint_to_skins.entry(joint) {
+            entry.get_mut().retain(|skin| *skin != skinned_mesh_entity);
+            if entry.get_mut().is_empty() {
+                entry.remove();
             }
         }
+    }
+
+    // Update the total number of joints.
+    skin_uniforms.total_joints -= old_skin_uniform_info.joints.len();
+}
+
+/// Marks [`SkinnedMesh`] components as changed if the
+/// [`SkinnedMeshInverseBindposes`] assets that they refer to changed.
+///
+/// This is needed to ensure that we pick up the new inverse bindposes in
+/// [`extract_skins`].
+pub fn mark_skins_as_changed_if_their_inverse_bindposes_changed(
+    mut events: EventReader<AssetEvent<SkinnedMeshInverseBindposes>>,
```

Comment:
If you impl `AsAssetId` for `SinnedMesh` , you can use the new `AssetChanged<SkinnedMesh>` filter. Since you're already doing the table scan, this should be pretty much equivalent in performance, I believe.

Reply:
Nice! Made that change.